File: kmp/composeApp/src/commonTest/kotlin/com/tanyudii/tmuxweb/domain/TerminalClipboardTest.kt
package com.tanyudii.tmuxweb.domain

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

/** Ports public/terminal-clipboard.test.js's cases 1:1 for isCopyShortcut/copyResultMessage. */
class TerminalClipboardTest {
    @Test
    fun `recognizes Cmd+C on keydown`() {
        val result = isCopyShortcut(type = "keydown", metaKey = true, ctrlKey = false, shiftKey = false, key = "c")

        assertTrue(result)
    }

    @Test
    fun `is case-insensitive on the key value`() {
        val result = isCopyShortcut(type = "keydown", metaKey = true, ctrlKey = false, shiftKey = false, key = "C")

        assertTrue(result)
    }

    @Test
    fun `ignores keyup so the copy doesn't fire twice per press`() {
        val result = isCopyShortcut(type = "keyup", metaKey = true, ctrlKey = false, shiftKey = false, key = "c")

        assertFalse(result)
    }

    @Test
    fun `recognizes plain Ctrl+C as a copy attempt, same convention as coolify's terminal`() {
        val result = isCopyShortcut(type = "keydown", metaKey = false, ctrlKey = true, shiftKey = false, key = "c")

        assertTrue(result)
    }

    @Test
    fun `ignores unrelated Cmd shortcuts`() {
        val result = isCopyShortcut(type = "keydown", metaKey = true, ctrlKey = false, shiftKey = false, key = "v")

        assertFalse(result)
    }

    @Test
    fun `ignores Cmd+Shift+C so it doesn't collide with the macOS devtools inspector shortcut`() {
        val result = isCopyShortcut(type = "keydown", metaKey = true, ctrlKey = false, shiftKey = true, key = "C")

        assertFalse(result)
    }

    @Test
    fun `ignores Ctrl+Shift+C so it doesn't collide with the Windows Linux devtools inspector shortcut`() {
        val result = isCopyShortcut(type = "keydown", metaKey = false, ctrlKey = true, shiftKey = true, key = "C")

        assertFalse(result)
    }

    @Test
    fun `success message confirms the copy`() {
        assertEquals("Copied", copyResultMessage(success = true))
    }

    @Test
    fun `failure message tells the user to copy manually`() {
        assertEquals("Auto-copy failed — select the text and copy manually", copyResultMessage(success = false))
    }

    @Test
    fun `no-selection message tells the user how to select terminal text`() {
        assertEquals(
            "No text selected — hold Option (Mac) or Shift (Windows/Linux) while dragging, then copy again",
            COPY_NO_SELECTION_MESSAGE,
        )
    }
}

File: web/src/domain/terminalClipboard.test.ts
import { describe, expect, test } from "vitest";
import { COPY_NO_SELECTION_MESSAGE, copyResultMessage, isCopyShortcut } from "./terminalClipboard";

// Ports kmp/.../domain/TerminalClipboardTest.kt 1:1 (itself a port of
// public/terminal-clipboard.test.js).
describe("isCopyShortcut", () => {
  test("recognizes Cmd+C on keydown", () => {
    const result = isCopyShortcut({ type: "keydown", metaKey: true, ctrlKey: false, shiftKey: false, key: "c" });

    expect(result).toBe(true);
  });

  test("is case-insensitive on the key value", () => {
    const result = isCopyShortcut({ type: "keydown", metaKey: true, ctrlKey: false, shiftKey: false, key: "C" });

    expect(result).toBe(true);
  });

  test("ignores keyup so the copy doesn't fire twice per press", () => {
    const result = isCopyShortcut({ type: "keyup", metaKey: true, ctrlKey: false, shiftKey: false, key: "c" });

    expect(result).toBe(false);
  });

  test("recognizes plain Ctrl+C as a copy attempt -- same convention as coolify's terminal", () => {
    const result = isCopyShortcut({ type: "keydown", metaKey: false, ctrlKey: true, shiftKey: false, key: "c" });

    expect(result).toBe(true);
  });

  test("ignores unrelated Cmd shortcuts", () => {
    const result = isCopyShortcut({ type: "keydown", metaKey: true, ctrlKey: false, shiftKey: false, key: "v" });

    expect(result).toBe(false);
  });

  test("ignores Cmd+Shift+C so it doesn't collide with the macOS devtools inspector shortcut", () => {
    const result = isCopyShortcut({ type: "keydown", metaKey: true, ctrlKey: false, shiftKey: true, key: "C" });

    expect(result).toBe(false);
  });

  test("ignores Ctrl+Shift+C so it doesn't collide with the Windows/Linux devtools inspector shortcut", () => {
    const result = isCopyShortcut({ type: "keydown", metaKey: false, ctrlKey: true, shiftKey: true, key: "C" });

    expect(result).toBe(false);
  });
});

describe("copyResultMessage", () => {
  test("success message confirms the copy", () => {
    expect(copyResultMessage(true)).toBe("Copied");
  });

  test("failure message tells the user to copy manually", () => {
    expect(copyResultMessage(false)).toBe("Auto-copy failed — select the text and copy manually");
  });
});

test("no-selection message tells the user how to select terminal text", () => {
  expect(COPY_NO_SELECTION_MESSAGE).toBe(
    "No text selected — hold Option (Mac) or Shift (Windows/Linux) while dragging, then copy again",
  );
});
